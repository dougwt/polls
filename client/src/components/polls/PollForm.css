.btn-delete {
  margin-left: 2em;
}
.btn-delete:hover {
  background-color: #ef476f !important;
  color: #fcfcfc !important;
}
.btn-small {
  height: 32.4px;
  line-height: 32.4px;
  font-size: 13px;
}
.red-text {
  color: #ef476f !important;
}
