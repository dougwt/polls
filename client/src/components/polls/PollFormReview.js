import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Card, Input, Button, Row, Icon } from 'react-materialize';
import * as actions from '../../actions';

export const PollFormReview = ({ onCancel, formValues, submitPoll, history }) => {
  return (
    <div>
      <h5>Please confirm your poll</h5>

      {/* {console.log(formValues)} */}

      <Row>
    		<Card className='darken-1' title={formValues.question}>

          {/* TODO: add spacing here via CSS */}

          <div className="choices">
            {renderChoices(formValues)}
          </div>

          <Row className="center-align">
            <Button
              className="teal"
              disabled
            >
              Vote
            </Button>
          </Row>

        </Card>
      </Row>

      <Row>
        <Button
          node="a"
          className="red btn-back btn-flat white-text"
          waves="light"
          onClick={onCancel}
        >
          Back
          <Icon left>keyboard_arrow_left</Icon>
        </Button>

        <Button
          className="teal btn-create btn-flat right white-text"
          waves="light"
          onClick={() => submitPoll(formValues, history)}
        >
          Create
          <Icon right>create</Icon>
        </Button>
      </Row>
    </div>
  )
};

function renderChoices(values) {
  let choices = [];

  for (let choice = 1; choice <= values.choices.length; choice++) {
    let field = `choice_${choice}`;
    if(values[field]) {
      choices.push(
        <Row key={choice}>
          <Input name="choices" type="radio" value={choice.toString()} label={values[field]} />
        </Row>
      )
    }
  }

  return choices;
};

function mapStateToProps(state) {
  return {
    formValues: state.form.pollNewForm.values
  };
}

export default connect(mapStateToProps, actions)(withRouter(PollFormReview));
