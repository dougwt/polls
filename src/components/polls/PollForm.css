.input-field .prefix.active {
  color: #06bad6;
}

input:not([type]):focus:not([readonly]),
input[type='text']:not(.browser-default):focus:not([readonly]),
input[type='password']:not(.browser-default):focus:not([readonly]),
input[type='email']:not(.browser-default):focus:not([readonly]),
input[type='url']:not(.browser-default):focus:not([readonly]),
input[type='time']:not(.browser-default):focus:not([readonly]),
input[type='date']:not(.browser-default):focus:not([readonly]),
input[type='datetime']:not(.browser-default):focus:not([readonly]),
input[type='datetime-local']:not(.browser-default):focus:not([readonly]),
input[type='tel']:not(.browser-default):focus:not([readonly]),
input[type='number']:not(.browser-default):focus:not([readonly]),
input[type='search']:not(.browser-default):focus:not([readonly]),
textarea.materialize-textarea:focus:not([readonly]) {
  border-bottom: 1px solid #06bad6;
  -webkit-box-shadow: 0 1px 0 0 #06bad6;
  box-shadow: 0 1px 0 0 #06bad6;
}
input.valid:not([type]),
input.valid:not([type]):focus,
input[type='text'].valid:not(.browser-default),
input[type='text'].valid:not(.browser-default):focus,
input[type='password'].valid:not(.browser-default),
input[type='password'].valid:not(.browser-default):focus,
input[type='email'].valid:not(.browser-default),
input[type='email'].valid:not(.browser-default):focus,
input[type='url'].valid:not(.browser-default),
input[type='url'].valid:not(.browser-default):focus,
input[type='time'].valid:not(.browser-default),
input[type='time'].valid:not(.browser-default):focus,
input[type='date'].valid:not(.browser-default),
input[type='date'].valid:not(.browser-default):focus,
input[type='datetime'].valid:not(.browser-default),
input[type='datetime'].valid:not(.browser-default):focus,
input[type='datetime-local'].valid:not(.browser-default),
input[type='datetime-local'].valid:not(.browser-default):focus,
input[type='tel'].valid:not(.browser-default),
input[type='tel'].valid:not(.browser-default):focus,
input[type='number'].valid:not(.browser-default),
input[type='number'].valid:not(.browser-default):focus,
input[type='search'].valid:not(.browser-default),
input[type='search'].valid:not(.browser-default):focus,
textarea.materialize-textarea.valid,
textarea.materialize-textarea.valid:focus,
.select-wrapper.valid > input.select-dropdown {
  border-bottom: 1px solid #06d6a0;
  -webkit-box-shadow: 0 1px 0 0 #06d6a0;
  box-shadow: 0 1px 0 0 #06d6a0;
}
input:not([type]):focus:not([readonly]) + label,
input[type='text']:not(.browser-default):focus:not([readonly]) + label,
input[type='password']:not(.browser-default):focus:not([readonly]) + label,
input[type='email']:not(.browser-default):focus:not([readonly]) + label,
input[type='url']:not(.browser-default):focus:not([readonly]) + label,
input[type='time']:not(.browser-default):focus:not([readonly]) + label,
input[type='date']:not(.browser-default):focus:not([readonly]) + label,
input[type='datetime']:not(.browser-default):focus:not([readonly]) + label,
input[type='datetime-local']:not(.browser-default):focus:not([readonly])
  + label,
input[type='tel']:not(.browser-default):focus:not([readonly]) + label,
input[type='number']:not(.browser-default):focus:not([readonly]) + label,
input[type='search']:not(.browser-default):focus:not([readonly]) + label,
textarea.materialize-textarea:focus:not([readonly]) + label {
  color: #06bad6;
}
.btn-delete {
  margin-left: 2em;
}
.btn-delete:hover {
  background-color: #ef476f !important;
  color: #fcfcfc !important;
}
.btn-add-choice:hover,
.btn-remove-choice:hover {
  background-color: #06bad6 !important;
}
.btn-small {
  height: 32.4px;
  line-height: 32.4px;
  font-size: 13px;
}
.red-text {
  color: #ef476f !important;
}
