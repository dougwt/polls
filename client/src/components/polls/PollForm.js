import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { reduxForm, Field, FieldArray } from 'redux-form';
import { Button, Row, Icon, ProgressBar } from 'react-materialize';
import PollField from './PollField';
import './PollForm.css';

export class PollForm extends Component {
  componentDidMount() {
    if (this.props.onReset) {
      this.props.onReset();
    }
  }

  render() {
    return (
      <div>
        <form onSubmit={this.props.handleSubmit(this.props.onSubmit)}>
          <Row>
            <Field
              label="What question would you like to ask?"
              icon="question_answer"
              name="question"
              component={PollField}
            />
          </Row>

          <Row>
            <FieldArray
              name="choices"
              component={this.renderChoices.bind(this)}
            />
          </Row>

          {this.renderError() ? this.renderError() : this.renderSpinner()}

          <Row>
            <Button
              className="btn btn-back red white-text"
              onClick={this.props.onCancel}
              type="button"
            >
              Cancel
              <Icon left>close</Icon>
            </Button>

            {this.renderAsyncButton()}

            <Button className="btn-next teal right white-text" waves="light">
              Next
              <Icon right>keyboard_arrow_right</Icon>
            </Button>
          </Row>
        </form>
      </div>
    );
  }

  renderChoices({ fields, meta: { error } }) {
    function addField(fields) {
      if (fields.length <= 5) {
        return fields.push();
      }
    }

    function removeField(fields) {
      if (fields.length >= 3) {
        return fields.pop();
      }
    }

    function renderChoiceField(index) {
      const choiceIcons = {
        1: 'looks_one',
        2: 'looks_two',
        3: 'looks_3',
        4: 'looks_4',
        5: 'looks_5',
        6: 'looks_6',
        7: 'looks_7',
        8: 'looks_8',
        9: 'looks_9'
      };

      return (
        <Field
          label={`Choice ${index}`}
          icon={choiceIcons[index]}
          name={`choice_${index}`}
          key={index}
          component={PollField}
        />
      );
    }

    function renderChoiceError(error) {
      if (error) {
        return <li className="center-align red-text">{error}</li>;
      }
    }

    return (
      <ul>
        {fields.map((choice, index) => (
          <li key={index}>{renderChoiceField(index + 1)}</li>
        ))}
        {renderChoiceError(error)}

        <li className="center-align">
          <Button
            flat
            className={
              'btn-add-choice' + (fields.length > 5 ? ' disabled' : '')
            }
            onClick={() => addField(fields)}
            node="a"
          >
            <Icon left>add</Icon>
            Add Choice
          </Button>
          <Button
            flat
            className={
              'btn-remove-choice' + (fields.length < 3 ? ' disabled' : '')
            }
            onClick={() => removeField(fields)}
            node="a"
          >
            <Icon left>delete</Icon>
            Remove Choice
          </Button>
        </li>
      </ul>
    );
  }

  renderError() {
    if (this.props.error) {
      return (
        <Row className="red-text center-align error">
          <strong>Error:</strong> {this.props.error.message}
        </Row>
      );
    }
  }

  renderSpinner() {
    if (this.props.waiting) {
      return (
        <Row>
          <ProgressBar />
        </Row>
      );
    }

    return <Row>&nbsp;</Row>;
  }

  renderAsyncButton() {
    if (!this.props.onDelete) {
      return;
    }

    if (this.props.waiting) {
      return (
        <Button
          className="btn disabled btn-small btn-delete btn-flat btn-red red-text"
          onClick={this.props.onDelete}
          type="button"
        >
          Deleting...
          <Icon left>delete</Icon>
        </Button>
      );
    }

    return (
      <Button
        className="btn btn-small btn-delete btn-flat btn-red red-text"
        onClick={this.props.onDelete}
        type="button"
      >
        Delete Poll
        <Icon left>delete</Icon>
      </Button>
    );
  }
}
PollForm.propTypes = {
  error: PropTypes.object,
  handleSubmit: PropTypes.func.isRequired,
  initialValues: PropTypes.object,
  onReset: PropTypes.func,
  onCancel: PropTypes.func.isRequired,
  onDelete: PropTypes.func,
  onSubmit: PropTypes.func.isRequired,
  waiting: PropTypes.bool.isRequired
};

function validate(values) {
  const errors = {};

  if (!values.question) {
    errors.question = 'You must provide a poll question';
  }

  if (!values.choices || values.choices.length < 2) {
    errors.choices = { _error: 'At least two choices must be entered' };
  } else if (values.choices.length > 6) {
    errors.choices = { _error: 'No more than six choices allowed' };
  }

  for (
    let choice = 1;
    choice <= (values.choices ? values.choices.length : 0);
    choice++
  ) {
    let field = `choice_${choice}`;
    if (!values[field]) {
      errors[field] = `You must provide text for Choice ${choice}`;
    }
  }

  return errors;
}

function mapStateToProps(state) {
  return {
    error: state.poll.error,
    waiting: state.poll.waiting
  };
}

export default compose(
  reduxForm({
    validate,
    form: 'pollForm',
    destroyOnUnmount: false,
    enableReinitialize: true // allow form to re-init b/w Edit & New instances
  }),
  connect(mapStateToProps)
)(PollForm);
