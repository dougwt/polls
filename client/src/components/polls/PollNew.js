import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import requireAuth from '../requireAuth';
import PollForm from './PollForm';
import PollFormReview from './PollFormReview';
import * as actions from '../../actions';

const initialValues = { choices: [0, 1] };

export class PollNew extends Component {
  constructor(props) {
    super(props);
    this.state = { showReview: false };
  }

  renderContent() {
    if (this.state.showReview) {
      return (
        <PollFormReview
          onCancel={() => this.setState({ showReview: false })}
          onSave={() => {
            const poll = {
              question: this.props.formValues.question,
              choices: this.props.formValues.choices.map((choice, i) => {
                return {
                  text: this.props.formValues[`choice_${i + 1}`],
                  votes: 0
                };
              })
            };

            if (this.props.auth) {
              poll.owner = this.props.auth._id;
            }
            this.props.createPoll(poll, () => {
              this.props.history.push('/polls');
            });
          }}
        />
      );
    }

    return (
      <PollForm
        onCancel={() => this.props.history.push('/polls')}
        onSubmit={() => this.setState({ showReview: true })}
        initialValues={initialValues}
      />
    );
  }

  render() {
    return (
      <div className="section">
        <div className="container">
          <div className="row">
            <h3 className="header section-title">Create a Poll</h3>
            {this.renderContent()}
          </div>
        </div>
      </div>
    );
  }
}
PollNew.propTypes = {
  createPoll: PropTypes.func.isRequired,
  auth: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]),
  history: PropTypes.object.isRequired,
  formValues: PropTypes.object.isRequired
};

function mapStateToProps(state) {
  return {
    auth: state.auth,
    formValues: (state.form.pollForm && state.form.pollForm.values) || {}
  };
}

export default compose(
  connect(mapStateToProps, actions),
  requireAuth,
  withRouter
)(PollNew);
