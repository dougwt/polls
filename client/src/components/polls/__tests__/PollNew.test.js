import React from 'react';
import { shallow } from 'enzyme';
import { PollNew } from '../PollNew';
import PollForm from '../PollForm';
import PollFormReview from '../PollFormReview';

describe('PollNew', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      auth: {},
      resetCreatePoll: jest.fn(),
      createPoll: jest.fn(),
      formValues: {},
      history: {}
    };

    wrapper = shallow(<PollNew {...props} />);
  });

  it('renders properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('when the form is initially loaded', () => {
    it('contains the proper component state', () => {
      expect(wrapper.state('showReview')).toEqual(false);
    });

    it('shows the PollForm component', () => {
      expect(wrapper.find(PollForm).length).toEqual(1);
    });

    it('does not show the PollFormReview component', () => {
      expect(wrapper.find(PollFormReview).length).toEqual(0);
    });
  });

  describe('when the form has been submitted for review', () => {
    beforeEach(() => {
      wrapper.setState({ showReview: true });
    });

    it('shows the PollFormReview component', () => {
      expect(wrapper.find(PollFormReview).length).toEqual(1);
    });

    it('does not show the PollForm component', () => {
      expect(wrapper.find(PollForm).length).toEqual(0);
    });
  });
});
