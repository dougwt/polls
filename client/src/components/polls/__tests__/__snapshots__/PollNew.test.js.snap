// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollNew renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PollNew
    auth={Object {}}
    createPoll={[MockFunction]}
    formValues={Object {}}
    history={Object {}}
    resetCreatePoll={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="container"
      >
        <div
          className="row"
        >
          <h3
            className="header section-title"
          >
            Create a Poll
          </h3>
          <ReduxForm
            initialValues={
              Object {
                "choices": Array [
                  0,
                  1,
                ],
              }
            }
            onCancel={[Function]}
            onReset={[MockFunction]}
            onSubmit={[Function]}
          />
        </div>
      </div>,
      "className": "section",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="row"
        >
          <h3
            className="header section-title"
          >
            Create a Poll
          </h3>
          <ReduxForm
            initialValues={
              Object {
                "choices": Array [
                  0,
                  1,
                ],
              }
            }
            onCancel={[Function]}
            onReset={[MockFunction]}
            onSubmit={[Function]}
          />
        </div>,
        "className": "container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3
              className="header section-title"
            >
              Create a Poll
            </h3>,
            <ReduxForm
              initialValues={
                Object {
                  "choices": Array [
                    0,
                    1,
                  ],
                }
              }
              onCancel={[Function]}
              onReset={[MockFunction]}
              onSubmit={[Function]}
            />,
          ],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Create a Poll",
              "className": "header section-title",
            },
            "ref": null,
            "rendered": "Create a Poll",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "initialValues": Object {
                "choices": Array [
                  0,
                  1,
                ],
              },
              "onCancel": [Function],
              "onReset": [MockFunction],
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="container"
        >
          <div
            className="row"
          >
            <h3
              className="header section-title"
            >
              Create a Poll
            </h3>
            <ReduxForm
              initialValues={
                Object {
                  "choices": Array [
                    0,
                    1,
                  ],
                }
              }
              onCancel={[Function]}
              onReset={[MockFunction]}
              onSubmit={[Function]}
            />
          </div>
        </div>,
        "className": "section",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="row"
          >
            <h3
              className="header section-title"
            >
              Create a Poll
            </h3>
            <ReduxForm
              initialValues={
                Object {
                  "choices": Array [
                    0,
                    1,
                  ],
                }
              }
              onCancel={[Function]}
              onReset={[MockFunction]}
              onSubmit={[Function]}
            />
          </div>,
          "className": "container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3
                className="header section-title"
              >
                Create a Poll
              </h3>,
              <ReduxForm
                initialValues={
                  Object {
                    "choices": Array [
                      0,
                      1,
                    ],
                  }
                }
                onCancel={[Function]}
                onReset={[MockFunction]}
                onSubmit={[Function]}
              />,
            ],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Create a Poll",
                "className": "header section-title",
              },
              "ref": null,
              "rendered": "Create a Poll",
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "initialValues": Object {
                  "choices": Array [
                    0,
                    1,
                  ],
                },
                "onCancel": [Function],
                "onReset": [MockFunction],
                "onSubmit": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
